### PR TITLE
Simple global load balancer proposal

### PR DESCRIPTION
**Pull request by raffaelespazzoli**
As discussed I am sending a KEP for the simple global load balancer proposal.
This is related to #2136

### CODE REVIEW

File: design/simple-global-load-balancing.md [hunk 1]
```diff
@@ -0,0 +1,938 @@
...
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+VIPs assigned via DNS will still belong to an address pool. A series of check will occur before the IP returned by the DNS will be actually assigned. At a minium the following checks:
+- the IP belongs to allowed CIDRs
+- the IP is available.
+
+A possible best practice to mitigate the risk of failure of these checks is to dedicate an IPaddress pool just ofr the globally load balanced services. This way the pool usage (taken IPs vs free IPs) will be the same across the clusters.
```
Comment (fedepaol) [inline design/simple-global-load-balancing.md:355]: This must be considered. In the version I am proposing (the ips go to the status sub structure), we won't notice conflicts until they happen. What I think it makes sense is to give IPs assigned manually precedence (i.e. if we have two IPAddressPools, where the IP belongs to one because of a manually assigned cidr, and to the other because of a dns source, the manual one will take precedence and we'll ignore the dns one).

File: design/simple-global-load-balancing.md [hunk 2]
```diff
@@ -0,0 +1,938 @@
...
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+-->
+
+- <test>: <link to test coverage>
+
+##### e2e tests
+
+<!--
```
Comment (fedepaol) [inline design/simple-global-load-balancing.md:456]: We should describe which tests we want to add and how (a test plan).

File: design/simple-global-load-balancing.md [hunk 3]
```diff
@@ -0,0 +1,938 @@
...
+The canonical place for the latest set of instructions (and the likely source
+of this file) is [here](/keps/NNNN-kep-template/README.md).
+
+**Note:** Any PRs to move a KEP to `implementable`, or significant changes once
+it is marked `implementable`, must be approved by each of the KEP approvers.
+If none of those approvers are still appropriate, then changes to that list
+should be approved by the remaining approvers and/or the owning SIG (or
+SIG Architecture for cross-cutting KEPs).
+-->
+# MLB-0002: Simple Global Load Balancing support
```
Comment (fedepaol) [inline design/simple-global-load-balancing.md:61]: I'd reword this to "LoadBalancerIP DNS source" or something along that line

**Comment by fedepaol** [comment]
cc @oribon

**Comment by oribon** [comment]
thanks for opening this!
I think having addresses allocated from an external source (be it dns/dhcp/..) is interesting and can open up a lot of new use-cases.
I agree with @fedepaol that this should be more focused on "setting addresses via dns" rather than on the specific use-case (global load balancers), and that it shouldn't be configured with a new annotation as we're trying to have a clear API with the CRDs.
The proposed solution needs to be generic enough to enable different external sources and it doesn't have to be embedded in the current process as this can easily be a different controller that syncs IPAddressPools.
I don't really like the idea of modifying the current IPAddressPool API (at a glance) as today it's clear what are its responsibilities and boundaries and I wouldn't want to over-complicate it.
I think something in the spirit of [external secrets](https://external-secrets.io/latest/introduction/overview/) might be something worth exploring to enable this, where we have a new API that expresses how IPAddressPools can be generated by an external source.
wdyt?

File: design/simple-global-load-balancing.md [hunk 4]
```diff
@@ -0,0 +1,938 @@
+<!--
```
[the remark this answers is not in the document]
Comment (raffaelespazzoli) [reply design/simple-global-load-balancing.md:1]: done

File: design/simple-global-load-balancing.md [hunk 5]
```diff
@@ -0,0 +1,938 @@
...
+
+That said at the moment there is no simple way to coordinate different instances of Metallb to pick the sam VIP for the same LoadBalancer service when deployed across different clusters.
+
+So some coordination is needed between the clusters across which one needs to establish global load balancing. And yet the metallb operator is cluster-bound and cannot "see" beyond the cluster it's running on.
+
+We propose to use DNS as a means of coordination between the various metallb instances.
+
+It should be possible to express the desire to have the VIP of a LoadBalancer service assigned by doing a dns lookup with a given FQDN.
+
+For example if a LoadBalancer service is annotated with: `metallb.io/assign-from-fqdn: myglobal.service.example.io`
```
[the remark this answers is not in the document]
Comment (raffaelespazzoli) [reply design/simple-global-load-balancing.md:239]: If I understand Metalb correclty, `IPaddressPool` is a highly powerful API which you wouldn't give access to to the developer. But the fqdn from which the IP needs to be takes is developer input. 
So I don;t think this is a good approach. The main requirement is that as a developer I should be able to define my globally loadbalanced services

File: design/simple-global-load-balancing.md [hunk 6]
```diff
@@ -0,0 +1,938 @@
...
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+VIPs assigned via DNS will still belong to an address pool. A series of check will occur before the IP returned by the DNS will be actually assigned. At a minium the following checks:
+- the IP belongs to allowed CIDRs
+- the IP is available.
+
+A possible best practice to mitigate the risk of failure of these checks is to dedicate an IPaddress pool just ofr the globally load balanced services. This way the pool usage (taken IPs vs free IPs) will be the same across the clusters.
+
+Also because it's possible that the DNS entry changes, the controller will have to check every so often with the DNS to ensure that the IP is still correct. The check frequency should be driven by the DNS record TTL.
```
[the remark this answers is not in the document]
Comment (raffaelespazzoli) [reply design/simple-global-load-balancing.md:357]: we could.

File: design/simple-global-load-balancing.md (hunk 3 above)
Comment (raffaelespazzoli) [reply design/simple-global-load-balancing.md:61]: done

File: design/simple-global-load-balancing.md (hunk 2 above)
Comment (raffaelespazzoli) [reply design/simple-global-load-balancing.md:456]: I am not sure I am the right person to do that not being too familiar with the code, but I could  give it a try

File: design/simple-global-load-balancing.md [hunk 7]
```diff
@@ -0,0 +1,938 @@
...
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+Metalb is designed to be a create and manage local load balancer. But the BGP protocol can be used to create global load balancer using the concept of equal cost multi path (ECMP).
```
[the remark this answers is not in the document]
Comment (fedepaol) [reply design/simple-global-load-balancing.md:21]: I'd reword this part too

File: design/simple-global-load-balancing.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (fedepaol) [reply design/simple-global-load-balancing.md:58]: I agree that you can't give the IPAddressPool to the user, but it's for the same reason you can't give him the annotation: it will give the tenant of the cluster the ability to pick whatever ip he wants. Also, as you described, this must be agreed with the cluster admin in any case, as the ip will have to match one available in the IPAddressPool. In general, the available ips are under the admin's control.

So the right workflow has to be "I have this service, I want to tie it to this fqdn, please admin configure me to do that".
As I wrote, we need to move the responsability of the ip assignement away from the service, otherwise it'd be easy to abuse of the feature and steal ips to other tenants.

[the remark this answers is not in the document]
Comment (raffaelespazzoli) [reply design/simple-global-load-balancing.md:58]: The FQDN must be user input. Also we cannot introduce a process by which the user has to ask something to the platform engineer. It would be like if the user had to ask the platform engineering to create an Ingress after the chose the fqdn. The process has to be completely self-serviceable.
`but it's for the same reason you can't give him the annotation: it will give the tenant of the cluster the ability to pick whatever ip he wants`: how do you reach that conclusion. 
As a user I request the VIPs behind a given FQDN, the system will accept that request only after passing the checks described in the doc. Otherwise the system will reject the request. Exactly as it happens with Ingresses if a user requests an Ingress with an FQDN that is already claimed.

[the remark this answers is not in the document]
Comment (fedepaol) [reply design/simple-global-load-balancing.md:58]: > The FQDN must be user input. Also we cannot introduce a process by which the user has to ask something to the platform engineer. It would be like if the user had to ask the platform engineering to create an Ingress after the chose the fqdn. The process has to be completely self-serviceable. `but it's for the same reason you can't give him the annotation: it will give the tenant of the cluster the ability to pick whatever ip he wants`: how do you reach that conclusion. 

Because that fqdn can be associated to any ip (available, if we add the check), not only the one associated to the service. Additionally, if the user is asking for the IP associated to a given fqdn, then that IP must be registered (and kept in sync, some how), because the addresspools might go out of sync.

Having the IPAddressPool follow that IP in an automatic fashion removes part of the burden of having to change it when / if the dns entry changes. Once we have an IPAddressPool fed by the dns resolution, the user can ask to have that ip pinned to his service, or he can specify the pool annotation on the service to make sure the service is fed by that IP addresspool. This would also open the door to other ways to feed the ipaddresspools from different sources.

[the remark this answers is not in the document]
Comment (raffaelespazzoli) [reply design/simple-global-load-balancing.md:58]: I agree with the risk of the returned IP not fitting the associated address pool. And I addressed that risk in with the proposed checks in the caveat section.
To reiterate the concept, the proposed annotation is a hint to the IPAM part of the operator: given an AddressPool and an IP being requested, instead of the normal IPAM process, lookup the IP via DNS. After that, all the standard checks apply. Meaning if the IP is not in the CIDR range, or if it is already taken, the IP should not be assigned. In that case the Service should not receive an IP and an event should be emitted describing the problem.

[the remark this answers is not in the document]
Comment (itewk) [reply design/simple-global-load-balancing.md:58]: I agree that whatever the solution is here, it has to be self serviceable by the developers. any friction between developers owning their ideation to production workflow including full CI, continuous delivery, and continuous deployment, along with all associated deployment config, in this case including global load balancing config, isn't viable. The orgs that wins are the ones that give their developers the power, with the checks, to get to production without any, and i mean ZERO, in process approvals, rather the process itself is approved, and so long as you stay int he bounds you are good. what is an infra person but yet another person, there is no reason why these ip/fqdn decisions can't reside where they should, with the devs, with the ops teams providing the oversite.

additionally, tools like this shouldn't be imparting arbitrary restrictions based on what control models with hink clients want, rather, give them the options, then they can lock them down based on their policies. so, if you go with the annotation approach, or something like it, if client wants to lock that down, they can use other methods, such as admission controls, to prevent use. if the control is moved to a global CR that the devs have no chance of touch, we preclude those that desire to relinquish the control.

[the remark this answers is not in the document]
Comment (raffaelespazzoli) [reply design/simple-global-load-balancing.md:58]: @fedepaol I had an idea to perhaps meet the all requirements that we have.
What if we said that there is a controller that watches for  `metallb.io/assign-from-fqdn: myglobal.service.example.io` annotation. When this annotation is found the controller automatically creates an IPAddressPool like this:
```yaml
apiVersion: metallb.io/v1beta1
kind: IPAddressPool
metadata:
  name: example
  namespace: metallb-system
spec:
  sources:
      - type: dns
        fqdn: myglobal.service.example.io
```
And also assigns the service with that annotation pointing that pool to the Service: `metallb.universe.tf/address-pool: example`

This way we still the user to input the fqdn, and we use the IPAddressPool to assign the IP, like you wanted.

[the remark this answers is not in the document]
Comment (fedepaol) [reply design/simple-global-load-balancing.md:58]: I am more and more inclined to say this is something that should / can be handled by a controller that is external to MetalLB, as @oribon said when mentioning [external secrets](https://external-secrets.io/) with the mapping between the source (dns, dhcp) and the destination ipaddresspool defined in some CRD  (plus some property, like the service selector).

This would work and would also not add extra complexity to metallb, as it would be an external component. Given this, I think this must not be part of metallb but a sort of add-on. In the future, we might provide a reference one so users can write their own (and eventually contribute it), but this depends on our bandwith.

Also, this can be translated in your case with a controller that listens to the services, looks at the fqdn annotation to fetch (and reconciles) the ip.
When generating the ipaddresspool, this reconciler can add also a service selector so the addresspool is bound only to that service.
